packages/frontend/src/components/chat.tsx

```diff
@@ -45,37 +52,106 @@ function Message({ msg, from, peerId }: MessageProps) {
       console.log('gossipsub console log', evt.detail)
       // FIXME: Why does 'from' not exist on type 'Message'?
       const { topic, data } = evt.detail
+
+      switch (topic) {
+        case CHAT_TOPIC: {
+          chatMessageCB(evt, topic, data)
+          break
+        }
+        case CHAT_FILE_TOPIC: {
+          chatFileMessageCB(evt, topic, data)
+          break
+        }
+        default: {
+          throw new Error(`Unexpected gossipsub topic: ${topic}`)
+        }
+      }
+    }
+
+    const chatMessageCB = (evt: CustomEvent<Message>, topic: string, data: Uint8Array) => {
       const msg = new TextDecoder().decode(data)
       console.log(`${topic}: ${msg}`)
 
       // Append signed messages, otherwise discard
       if (evt.detail.type === 'signed') {
-        setMessageHistory([...messageHistory, { msg, from: 'other', peerId: evt.detail.from.toString() }])
+        setMessageHistory([...messageHistory, { msg, fileObjectUrl: undefined, from: 'other', peerId: evt.detail.from.toString() }])
+      }
+    }
+
+    const chatFileMessageCB = async (evt: CustomEvent<Message>, topic: string, data: Uint8Array) => {
+      const fileId = new TextDecoder().decode(data)
+
+      // if the message isn't signed, discard it.
+      if (evt.detail.type !== 'signed') {
+        return
       }
+      const senderPeerId = evt.detail.from;
+
+      const stream = await libp2p.dialProtocol(senderPeerId, FILE_EXCHANGE_PROTOCOL)
+      await pipe(
+        [uint8ArrayFromString(fileId)],
+        (source) => lp.encode(source),
+        stream,
+        (source) => lp.decode(source),
+        async function(source) {
+          for await (const data of source) {
+            const body: Uint8Array = data.subarray()
+            console.log(`request_response: response received: size:${body.length}`)
+
+            const msg: ChatMessage = {
+              msg: newChatFileMessage(fileId, body),
+              fileObjectUrl: window.URL.createObjectURL(new Blob([body])),
+              from: 'other',
+              peerId: senderPeerId.toString(),
+            }
+            setMessageHistory([...messageHistory, msg])
+          }
+        }
+      )
     }
 
     libp2p.services.pubsub.addEventListener('message', messageCB)
 
+    libp2p.handle(FILE_EXCHANGE_PROTOCOL, ({ stream }) => {
+      pipe(
+        stream.source,
+        (source) => lp.decode(source),
+        (source) => map(source, async (msg) => {
+          const fileId = uint8ArrayToString(msg.subarray())
+          const file = files.get(fileId)!
+          return file.body
+        }),
+        (source) => lp.encode(source),
+        stream.sink,
+      )
+    })
+
     return () => {
-      // Cleanup handlers 👇
-      // libp2p.pubsub.unsubscribe(CHAT_TOPIC)
-      libp2p.services.pubsub.removeEventListener('message', messageCB)
+      (async () => {
+        // Cleanup handlers 👇
+        // libp2p.pubsub.unsubscribe(CHAT_TOPIC)
```

Comment:
Could you remove the commented code please?

Reply:
thank you: 1f0bc87